plugins/optimization-detective/class-od-link-collection.php

```diff
@@ -255,9 +255,24 @@ public function get_response_header(): ?string {
 		$link_headers = array();
 
 		foreach ( $this->get_prepared_links() as $link ) {
-			// The about:blank is present since a Link without a reference-uri is invalid so any imagesrcset would otherwise not get downloaded.
-			$link['href'] = isset( $link['href'] ) ? esc_url_raw( $link['href'] ) : 'about:blank';
-			$link_header  = '<' . $link['href'] . '>';
+			if ( isset( $link['href'] ) ) {
+				$decoded_url = urldecode( $link['href'] );
+
+				// Encode characters not allowed in a URL per RFC 3986 (anything that is not among the reserved and unreserved characters).
+				$encoded_url  = preg_replace_callback(
+					'/[^A-Za-z0-9\-._~:\/?#\[\]@!$&\'()*+,;=%]/',
```

Comment:
```suggestion
					'/[^A-Za-z0-9\-._~:\/?#\[\]@!$&\'()*+,;=]/',
```

> - %: Percent sign %, used in percent-encoded characters in URLs.

Is a `%` correct here? The use of `urldecode()` should have already eliminated any % signs. Therefore, any remaining percent characters would be part of strings that aren't in an encoding, right? I added a failing test case that shows what should be expected but isn't currently the result: b1ed4c1cc77c7f7f3d9c3e75bdc8010e6fea8ce7

If the `%` is removed from this pattern, then the test is fixed.

Reply:
Right, completely missed that 😅

I've removed the % from the regex since urldecode() should handle any encoded characters.